### ISSUE TITLE
Change AWS NTP source to Amazon Time Sync Service

### ISSUE THREAD

**Issue by arlampin** [kubernetes/kops#4021]
Amazon has released Amazon Time Sync Service for AWS ([Blog post](https://aws.amazon.com/blogs/aws/keeping-time-with-amazon-time-sync-service/)). The docs have also been updated (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html).

I think it would be nice to have nodes in EC2 sync with EC2's own, internal NTP system, instead of using external NTP (Debian pool currently).

**Comment by gambol99** [comment]
/assign @gambol99

hi @arlampin ... you could use the [hooks](https://github.com/kubernetes/kops/blob/master/docs/cluster_spec.md#hooks) extension to get this configured, assuming it can't be done by a privileged daemonset.

**Comment by chrislovecnm** [comment]
We should check what to use in gce.

**Comment by chrislovecnm** [comment]
And vsphere and do should be left alone

**Comment by 007** [comment]
At a minimum `kops` should be substituting the `debian` pool NTP hosts with `amazon` pool hosts on AWS:
```
server 0.amazon.pool.ntp.org iburst 
server 1.amazon.pool.ntp.org iburst
server 2.amazon.pool.ntp.org iburst 
server 3.amazon.pool.ntp.org iburst
```

And similarly on GCE the `google` leap-smearing NTP host is probably the best thing to use:
```
server time.google.com iburst
```

Is that something that can/should be done by hooks, modifying the node FS / daemons? This feels different from the nvidia example, but I don't quite grok how it's altering the system kernel from within the hook-container - it looks like it's just installing all of the libraries and adding a shared mount to access them. It seems much more sensible to have NTP as part of the UserData, so it's done on install just like the rest of the node configuration.

Or is the recommendation to run this as something like a privileged CronJob, so it can still live on the kube side, but will affect the time for other containers on the host?

**Comment by chrislovecnm** [comment]
This should be a setup item completed by nodeup IMO.  @007 agree with all of the above :)

**Comment by 007** [comment]
/remove-lifecycle stale

Also ref https://github.com/kubernetes/kops/issues/4173, that can probably be closed as dupe of this?

**Comment by arlampin** [comment]
I would suggest using `169.254.169.123` on AWS as per [AWS user guide](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/set-time.html), instead of `amazon.pool.ntp.org`.

Also, GCE has `metadata.google` which provides similar functionality ([docs](https://cloud.google.com/compute/docs/instances/managing-instances#configure_ntp_for_your_instances))

**Comment by ddrugeon** [comment]
Hello there,
I am completely neebie to kops. I would also like to set NTP pool server to a dedicated one. I saw that I can do it thanks to specific hook. Can you give me an example how to proceed ?
Thanks,
David

**Comment by ddrugeon** [comment]
Here is my solution:

```
hooks:
  - name: change_ntp_server.service
    roles:
    - Node
    - Master
    after:
    - ntp.service
    manifest: |
      [Service]
      Type=oneshot
      RemainAfterExit=no
      ExecStart=/bin/sh -c "sed -i -- 's/pool/#pool/g' /etc/ntp.conf ; echo 'server 169.254.169.123 prefer iburst' >> /etc/ntp.conf"
      ExecStartPost=/bin/systemctl restart ntp.service

```

**Comment by 007** [comment]
/remove-lifecycle stale

Still valid as ever.

**Comment by 007** [comment]
Sorry I missed your birthday #4021 - happy birthday!

**Comment by ReillyProcentive** [comment]
/remove-lifecycle stale

Perhaps this could be taken care of after #6236 is merged?

**Comment by hairyhenderson** [comment]
#6236 has been merged now - any chance this can move forward?

**Comment by ReillyProcentive** [comment]
It's relatively simple to change this:
```
sudo sed -i "/#server/c\server 169.254.169.123 prefer iburst" /etc/ntp.conf
sudo sed -i "/^pool/d" /etc/ntp.conf
sudo systemctl restart ntpd.service
```
(from my custom image Packer job)

It wouldn't be terribly difficult to implement it into nodeup. Although, I wonder if using s[ystemd-timesyncd](https://wiki.archlinux.org/index.php/Systemd-timesyncd) wouldn't be the better option for operating systems that support it (which might be all of them actually).

**Comment by hairyhenderson** [comment]
ah! I didn't realize this was in the AMI, for some reason I thought nodeup already did this 😬

Thanks for your quick response, @ReillyProcentive!

**Comment by ReillyProcentive** [comment]
Nodeup installs ntp if it isn't already installed but uses the default configuration for it. In our case we're installing it ahead of time in a custom AMI and customizing ntp.conf to point to the AWS time service. Thus, when nodeup runs it just checks to see if ntp is present and if so makes sure it is enabled and started. This is our solution to nodeup not having native support for configuring the time source.

This issue is still valid as it could be an option in the kops spec to use cloud-specific time sources and have nodeup modify ntp.conf accordingly.

**Comment by nvanheuverzwijn** [comment]
Would fixing the ntp thing in https://github.com/kubernetes-sigs/image-builder/blob/master/images/kube-deploy/imagebuilder/templates/1.15-stretch.yml makes sense ?
The documentation about the AMI generation process is rather lacking :(

**Comment by ReillyProcentive** [comment]
Yes, it would make sense to add an AWS-specific command there to update the ntp config file.

**Comment by hairyhenderson** [comment]
Did https://github.com/kubernetes-sigs/image-builder/pull/127 fix this? /cc @qqshfox

**Comment by ari-becker** [comment]
Not fixed by https://github.com/kubernetes/kops/pull/8312 ?

**Comment by nvanheuverzwijn** [comment]
This issue is fixed. The latest onaster branch uses AWS NTP server.
